### PR TITLE
feat: Button Group component

### PR DESCRIPTION
**Pull request by spugachev**
### Description

Button Group component development

Rel: vAZbObs0rwNs

### How has this been tested?

* New unit- and integration tests
* New screenshot tests on permutations

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/button-group/internal.tsx [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+                <ItemElement
+                  item={itemOrGroup}
+                  onItemClick={onItemClick}
+                  ref={element => onSetButtonRef(itemOrGroup, element)}
+                />
+              );
+
+            const isGroupBefore = items[index - 1]?.type === 'group';
+            const isGroupNow = items[index]?.type === 'group';
+            const shouldAddDivider = isGroupBefore || (!isGroupBefore && isGroupNow && index !== 0);
```
Comment (pan-kot) [inline src/button-group/internal.tsx:212]: Let's add a dev warning (with warnOnce) for empty groups

File: src/button-group/internal.tsx [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        } else {
+          return [item.id];
+        }
+      };
+
+      const getAllIds = (item: ButtonGroupProps.Item) => {
+        if (item.type === 'icon-button') {
+          return [item.id];
+        } else {
+          const values = getAllIdsItemOrGroup(item);
```
Comment (pan-kot) [inline src/button-group/internal.tsx:71]: Why is it necessary to collect IDs of the menu items?

File: src/button-group/item-element.tsx [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+      return () => {
+        window.removeEventListener('pointerdown', handlePointerDownEvent);
+        window.removeEventListener('keydown', handleKeyDownEvent);
+        window.removeEventListener('btngroup-tooltip:toggle', handleTooltipToogleEvent as any);
+      };
+    }, [item.id, showTooltip]);
+
+    useEffect(() => {
+      window.dispatchEvent(
+        new CustomEvent('btngroup-tooltip:toggle', { detail: { open: showTooltip, trackKey: item.id } })
```
Comment (pan-kot) [inline src/button-group/item-element.tsx:101]: The custom event inside an effect logic is hard to follow and test against. Is it possible to make all checks in the click handler itself?

Instead of having tooltip show and tooltip feedback states here you can lift them to the button group level so that there is only one state for the entire component as there is only one tooltip that can be shown at a time. The state can be like tooltipItemId (can be null which means no tooltip or an item ID which defines for which item to show the tooltip) and tooltipFeedback (a boolean, when true - show feedback instead of text).

File: src/button-group/menu-dropdown-item.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      <ButtonDropdown
+        ref={ref}
+        variant="icon"
+        loading={item.loading}
+        loadingText={item.loadingText}
+        disabled={item.disabled}
+        items={item.items}
+        onItemClick={onClickHandler}
+        expandToViewport={expandToViewport}
+        ariaLabel={item.text}
```
Comment (pan-kot) [inline src/button-group/menu-dropdown-item.tsx:39]: When doing manual testing I noticed that text is not shown in a tooltip for menu buttons, could you verify please?

File: src/button-group/__tests__/icon-button-item.test.tsx [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  text: 'Test Button',
+  iconName: 'add-plus',
+  loading: false,
+  loadingText: 'Loading...',
+  disabled: false,
+  popoverFeedback: 'Action Popover',
+};
+
+describe('IconButtonItem', () => {
+  test('renders the button', () => {
```
[the remark this answers is not in the document]
Comment (spugachev) [reply src/button-group/__tests__/icon-button-item.test.tsx:26]: Removed

File: src/button-group/__tests__/menu-dropdown-item.test.tsx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+      items: [
+        { id: '5', text: 'Item 5' },
+        { id: '6', text: 'Item 6' },
+      ],
+    },
+  ],
+};
+
+describe('MenuDropdownItem', () => {
+  test('renders the button', () => {
```
[the remark this answers is not in the document]
Comment (spugachev) [reply src/button-group/__tests__/menu-dropdown-item.test.tsx:29]: Removed

File: src/button-group/icon-button-item.tsx [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    {
+      item,
+      onItemClick,
+    }: {
+      item: ButtonGroupProps.IconButton;
+      onItemClick?: (event: CustomEvent<ClickDetail>) => void;
+    },
+    ref: React.Ref<ButtonProps.Ref>
+  ) => {
+    const hasIcon = item.iconName || item.iconUrl || item.iconSvg;
```
[the remark this answers is not in the document]
Comment (spugachev) [reply src/button-group/icon-button-item.tsx:23]: Added

File: src/button-group/interfaces.ts [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+import { IconProps } from '../icon/interfaces';
+import { NonCancelableEventHandler } from '../internal/events';
+import { ButtonDropdownProps } from '../button-dropdown/interfaces';
+
+export interface ButtonGroupProps extends BaseComponentProps {
+  /**
+   * Adds `aria-label` to the button group toolbar element.
+   */
+  ariaLabel: string;
+  ariaLabelledby?: string;
```
[the remark this answers is not in the document]
Comment (spugachev) [reply src/button-group/interfaces.ts]: ariaLabelledby property is removed

File: src/button-group/interfaces.ts [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+     * Enabling this property will allow the dropdown to extend beyond that container by using fixed positioning and
+     * [React Portals](https://reactjs.org/docs/portals.html).
+     *
+     * Set this property if the dropdown would otherwise be constrained by a scrollable container,
+     * for example inside table and split view layouts.
+     *
+     * However, use discretion. We recommend you don't enable this property unless necessary
+     * because fixed positioning results in a slight, visible lag when scrolling complex pages.
+     */
+    expandToViewport?: boolean;
```
[the remark this answers is not in the document]
Comment (spugachev) [reply src/button-group/interfaces.ts]: Fixed

File: src/button-group/menu-dropdown-item.tsx (hunk 4 above)
Comment (spugachev) [reply src/button-group/menu-dropdown-item.tsx:39]: Currently it's not shown. When it is shown it looks weird. Let's discuss disign of this

File: src/button-group/item-element.tsx (hunk 3 above)
Comment (spugachev) [reply src/button-group/item-element.tsx:103]: The puppose of this code is to make synchronize toolpips beetween different ButtonGroup components on the same page. I agree that it's better to lift the code. The only question - should we keep global event taht closes other ButtonGroup's tooltips?

Comment (pan-kot) [reply src/button-group/item-element.tsx:103]: I don't see how the global event is useful. The tooltip of a particular button group instance must close when there is a click outside

Comment (spugachev) [reply src/button-group/item-element.tsx:103]: https://github.com/user-attachments/assets/d2d0f5c8-286c-471a-a73e-000828820f0f

Here is a demo

File: src/button-group/icon-button-item.tsx [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+    if (isDevelopment && !hasIcon) {
```
Comment (pan-kot) [inline src/button-group/icon-button-item.tsx:27]: I believe isDevelopment is already checked by warnOnce

Comment (pan-kot) [reply src/button-group/icon-button-item.tsx:27]: We should also add a unit test for the warning

File: src/button-group/interfaces.ts [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+export interface InternalButtonGroupProps extends ButtonGroupProps, InternalBaseComponentProps {}
+
+export namespace ButtonGroupProps {
+  // Fixes documenter output.
+  // When a union only has one element the resulting type for it is not string but object.
+  export type Variant = 'icon' | 'icon';
+
+  export type ItemOrGroup = Item | Group;
+  export type Item = IconButton | MenuDropdown;
```
Comment (pan-kot) [inline src/button-group/interfaces.ts:75]: Let's make the types Item and ItemOrGroup private (move outside the namespace)

File: src/button-group/item-element.tsx (hunk 3 above)
Comment (pan-kot) [reply src/button-group/item-element.tsx:103]: I think this one is probably fine - this is consistent with how our popover works so if we decide to fix that it is better to invent a holistic solution.

File: src/button-group/index.tsx [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+import useBaseComponent from '../internal/hooks/use-base-component';
+import { ButtonGroupProps } from './interfaces';
+import InternalButtonGroup from './internal';
+import { getBaseProps } from '../internal/base-component';
+
+export { ButtonGroupProps };
+
+const ButtonGroup = React.forwardRef((props: ButtonGroupProps, ref: React.Ref<ButtonGroupProps.Ref>) => {
+  const baseProps = getBaseProps(props);
+  const baseComponentProps = useBaseComponent('ButtonGroup');
```
[the remark this answers is not in the document]
Comment (pan-kot) [reply src/button-group/index.tsx]: Yes!

File: src/button-group/internal.tsx (hunk 1 above)
Comment (spugachev) [reply src/button-group/internal.tsx:252]: Added

File: src/button-group/internal.tsx (hunk 2 above)
Comment (spugachev) [reply src/button-group/internal.tsx]: Thes code was removed

File: src/button-group/interfaces.ts (hunk 11 above)
Comment (spugachev) [reply src/button-group/interfaces.ts:75]: Why do you want to do it? It ssems that users should be able to create items of these types

Comment (pan-kot) [reply src/button-group/interfaces.ts:75]: Ah, I forgot the type is used in the interface above. Nevermind - let's keep both Item and ItemOrGroup

File: src/button-group/internal.tsx [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+    const baseProps = getBaseProps(props);
+    const [focusedItemId, setFocusedItemId] = useState<null | string>(null);
+    const navigationAPI = useRef<SingleTabStopNavigationAPI>(null);
+    const containerObjectRef = useRef<HTMLDivElement>(null);
+    const containerRef = useMergeRefs(containerObjectRef, __internalRootRef);
+    const itemsRef = useRef<Record<string, ButtonProps.Ref | null>>({});
+    const itemWrappersRef = useRef<Record<string, HTMLDivElement | null>>({});
+    const [tooltipItemId, setTooltipItemId] = useState<string | null>(null);
+    const [isFeedbackTooltip, setIsFeedbackTooltip] = useState(false);
+    const allItems = useMemo(
```
Comment (pan-kot) [inline src/button-group/internal.tsx:48]: nit: I'd call the variable itemsById or similar to emphasise it is a dictionary type

File: src/button-group/internal.tsx [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+              <React.Fragment key={itemOrGroup.type === 'group' ? index : itemOrGroup.id}>
+                {shouldAddDivider && <div className={styles.divider} />}
+                {itemOrGroup.type === 'group' ? (
+                  <div key={index} role="group" aria-label={itemOrGroup.text} className={styles.group}>
+                    {itemOrGroup.items.map(item => (
+                      <React.Fragment key={item.id}>{itemContent(item)}</React.Fragment>
+                    ))}
+                  </div>
+                ) : (
+                  <React.Fragment key={itemOrGroup.id}>{itemContent(itemOrGroup)}</React.Fragment>
```
Comment (pan-kot) [inline src/button-group/internal.tsx:271]: This fragment is probably not needed here

File: src/button-group/item-element.tsx [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+      onItemClick,
+    }: {
+      item: ButtonGroupProps.Item;
+      dropdownExpandToViewport?: boolean;
+      onItemClick?: CancelableEventHandler<ButtonGroupProps.ItemClickDetails | ClickDetail>;
+    },
+    ref: React.Ref<HTMLDivElement>
+  ) => {
+    return (
+      <>
```
Comment (pan-kot) [inline src/button-group/item-element.tsx:23]: nit: you can remove the fragment in favour of having a switch or if/else in the component's function body

File: src/button-group/internal.tsx (hunk 13 above)
Comment (spugachev) [reply src/button-group/internal.tsx]: Renamed

File: src/button-group/internal.tsx (hunk 14 above)
Comment (spugachev) [reply src/button-group/internal.tsx]: Removed

File: src/button-group/item-element.tsx (hunk 15 above)
Comment (spugachev) [reply src/button-group/item-element.tsx]: Using switch now

File: src/button-group/internal.tsx [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+        onHome: () => focusElement(focusables[0]),
+        onEnd: () => focusElement(focusables[focusables.length - 1]),
+        onInlineStart: () => focusElement(focusables[circleIndex(activeIndex - 1, [0, focusables.length - 1])]),
+        onInlineEnd: () => focusElement(focusables[circleIndex(activeIndex + 1, [0, focusables.length - 1])]),
+      });
+    }
+
+    function focusElement(element: HTMLElement) {
+      setFeedbackItemId(null);
+      setTimeout(() => {
```
Comment (pan-kot) [inline src/button-group/internal.tsx:111]: what it the purpose of this timeout?

File: src/button-group/item-element.tsx [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+      };
+    }, [item.id, setFeedbackItemId, showTooltip]);
+
+    const onShowTooltip = (show: boolean) => {
+      if (feedbackItemId && (feedbackItemId !== item.id || (feedbackItemId && showTooltip))) {
+        return;
+      }
+
+      if (show) {
+        buttonRef.current?.focus();
```
Comment (pan-kot) [inline src/button-group/item-element.tsx:88]: This line steals the focus from the dropdown when it is open. The user moves focus to the menu button and presses Enter. The expectation is that the first item of the menu dropdown is focused but instead the focus comes back to the menu item.

Comment (pan-kot) [reply src/button-group/item-element.tsx:88]: @spugachev  I created a PR with a fix: https://github.com/cloudscape-design/components/pull/2491

File: src/button-group/icon-button-item.tsx (hunk 10 above)
Comment (spugachev) [reply src/button-group/icon-button-item.tsx]: How do you test warnOnce?

Comment (pan-kot) [reply src/button-group/icon-button-item.tsx]: We mock it and assert calls, see example: https://github.com/cloudscape-design/components/blob/2c3bfbe12bfe8773f12896a09664f911d468bd85/src/expandable-section/__tests__/expandable-section.test.tsx#L295

**Comment by pan-kot** [comment]
@spugachev I contributed a fix for tooltip conditions: https://github.com/cloudscape-design/components/pull/2496

File: src/button-group/interfaces.ts [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+    iconUrl?: string;
+    iconSvg?: React.ReactNode;
+    popoverFeedback?: React.ReactNode;
+  }
+
+  export interface MenuDropdown {
+    type: 'menu-dropdown';
+    id: string;
+    text: string;
+    disabled?: boolean;
```
Comment (johannes-weber) [inline src/button-group/interfaces.ts:99]: We recently introduced the disabled with reason feature (https://github.com/cloudscape-design/components/pull/2409). I can imagine we'll get some why they are not available on the ButtonGroup component.

Is there a specific reason we do not add the disabled reason here as well?

File: src/button-group/interfaces.ts [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+import { IconProps } from '../icon/interfaces';
+import { BaseComponentProps } from '../internal/base-component';
+import { NonCancelableEventHandler } from '../internal/events';
+import { InternalBaseComponentProps } from '../internal/hooks/use-base-component';
+
+export interface ButtonGroupProps extends BaseComponentProps {
+  /**
+   * Adds `aria-label` to the button group toolbar element.
+   */
+  ariaLabel: string;
```
[the remark this answers is not in the document]
Comment (pan-kot) [reply src/button-group/interfaces.ts:14]: Right, good catch 👍 

@spugachev , could you please update? Reference: 9Tj7AP2BUVuJ

File: src/button-group/interfaces.ts (hunk 18 above)
Comment (pan-kot) [reply src/button-group/interfaces.ts:99]: Yes, the disabled reasons are not supported for icon buttons as of yet.

**Comment by pan-kot** [comment]
Submitted a fix for the failing test and more tests: https://github.com/cloudscape-design/components/pull/2516

File: src/button-group/interfaces.ts [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+  /**
+   * Called when the user clicks on an item, and the item is not disabled. The event detail object contains the id of the clicked item.
+   */
+  onItemClick?: NonCancelableEventHandler<ButtonGroupProps.ItemClickDetails>;
+}
+
+export interface InternalButtonGroupProps extends ButtonGroupProps, InternalBaseComponentProps {}
+
+export namespace ButtonGroupProps {
+  // Fixes documenter output.
```
[the remark this answers is not in the document]
Comment (pan-kot) [reply src/button-group/interfaces.ts:74]: Good question. Here is the fix: https://github.com/cloudscape-design/documenter/pull/42

File: src/button-group/interfaces.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (pan-kot) [reply src/button-group/interfaces.ts]: Update: I fixed the property type, resolving

File: src/button-group/__tests__/button-group.test.tsx [hunk 21]
```diff
@@ -0,0 +1,253 @@
...
+    expect(wrapper.findMenuById('misc')!.getElement().getElementsByTagName('button')[0]).toHaveFocus();
+  });
+
+  test('focuses the correct item with keyboard', async () => {
+    const ref: { current: ButtonGroupProps.Ref | null } = { current: null };
+    const wrapper = renderButtonGroup({ items: items1 }, ref);
+    ref.current?.focus('copy');
+
+    fireEvent.keyDown(wrapper.getElement(), { keyCode: KeyCode.right });
+    await timeout();
```
Comment (johannes-weber) [inline src/button-group/__tests__/button-group.test.tsx:123]: Why do we need the timeouts? I assume it's because of waiting for the react updates. If so, why not using `act` instead?

File: src/button-group/icon-button-item.tsx [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    return (
+      <div ref={containerRef}>
+        <InternalButton
+          variant="icon"
+          loading={item.loading}
+          loadingText={item.loadingText}
+          disabled={item.disabled}
+          iconName={hasIcon ? item.iconName : 'close'}
```
Comment (johannes-weber) [inline src/button-group/icon-button-item.tsx:47]: What's the motivation behind using the `close` icon as default variant? I'd like to better understand why the `iconName` property is optional and not required.

File: src/button-group/interfaces.ts [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+
+import { ButtonDropdownProps } from '../button-dropdown/interfaces';
+import { IconProps } from '../icon/interfaces';
+import { BaseComponentProps } from '../internal/base-component';
+import { NonCancelableEventHandler } from '../internal/events';
+import { InternalBaseComponentProps } from '../internal/hooks/use-base-component';
+
+export interface ButtonGroupProps extends BaseComponentProps {
+  /**
+   * Adds `aria-label` to the button group toolbar element.
```
Comment (johannes-weber) [inline src/button-group/interfaces.ts:12]: Let's add a sentence about when to use this optional property as we have it in other components. Suggestion:
Use to assign unique labels when there are multiple button groups on one page.

File: src/button-group/item-element.tsx [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+    const onShowTooltipHard = (show: boolean) => {
+      if (!show && item.id !== tooltip?.item) {
+        return;
+      }
+
+      setTooltip(show ? { item: item.id, feedback: false } : null);
+    };
+
+    const onClickHandler = (event: CustomEvent<ButtonGroupProps.ItemClickDetails | ClickDetail>) => {
+      const popoverFeedback = 'popoverFeedback' in item && item.popoverFeedback;
```
Comment (johannes-weber) [inline src/button-group/item-element.tsx:86]: Minor naming suggestion:
```suggestion
      const hasPopoverFeedback = 'popoverFeedback' in item && item.popoverFeedback;
```

File: src/button-group/item-element.tsx [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+      tooltip,
+      setTooltip,
+      onItemClick,
+    }: {
+      item: ButtonGroupProps.Item;
+      dropdownExpandToViewport?: boolean;
+      tooltip: null | { item: string; feedback: boolean };
+      setTooltip: (tooltip: null | { item: string; feedback: boolean }) => void;
+      onItemClick?: NonCancelableEventHandler<ButtonGroupProps.ItemClickDetails> | undefined;
+    },
```
Comment (johannes-weber) [inline src/button-group/item-element.tsx:27]: Suggestion (readability): You could extract the prop definitions in a separtate `ItemElementProps` interface.

File: src/button-group/menu-dropdown-item.tsx [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+      item,
+      showTooltip,
+      onItemClick,
+      expandToViewport,
+    }: {
+      item: ButtonGroupProps.MenuDropdown;
+      showTooltip: boolean;
+      onItemClick?: CancelableEventHandler<ButtonGroupProps.ItemClickDetails>;
+      expandToViewport?: boolean;
+    },
```
Comment (johannes-weber) [inline src/button-group/menu-dropdown-item.tsx:28]: Suggestion (readability): You could extract the prop definitions in a separtate `MenuDropdownItemProps` interface.

File: src/button-group/menu-dropdown-item.tsx [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+                className={clsx(styles.tooltip, testUtilStyles['button-group-tooltip'])}
+              />
+            )}
+            <InternalButton
+              ref={triggerRef}
+              variant="icon"
+              ariaLabel={ariaLabel}
+              ariaExpanded={ariaExpanded}
+              className={clsx(styles.item, testUtilsClass)}
+              data-testid={item.id}
```
Comment (johannes-weber) [inline src/button-group/menu-dropdown-item.tsx:65]: We are setting the same `data-test-id` twice (in this line and above, in line 48) - can we avoid setting the same ID twice?

File: src/internal/hooks/use-focus-visible-state/index.ts [hunk 28]
```diff
@@ -0,0 +1,31 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import { useEffect, useState } from 'react';
+
+import { isModifierKey } from '@cloudscape-design/component-toolkit/internal';
+
+export function useFocusVisibleState() {
```
Comment (johannes-weber) [inline src/internal/hooks/use-focus-visible-state/index.ts:7]: I think we should add unit tests for this hook as it's now a standalone hook outside of the internal chart-plot component.

File: src/internal/context/single-tab-stop-navigation-context.tsx [hunk 29]
```diff
@@ -97,7 +97,7 @@ export const SingleTabStopNavigationProvider = forwardRef(
       const isUnregisteringFocusedNode = nodeBelongs(focusableElement, document.activeElement);
       if (isUnregisteringFocusedNode) {
         // Wait for unmounted node to get removed from the DOM.
-        requestAnimationFrame(() => onUnregisterActive?.(focusableElement));
+        setTimeout(() => onUnregisterActive?.(focusableElement), 0);
```
Comment (johannes-weber) [inline src/internal/context/single-tab-stop-navigation-context.tsx:100]: Why did we change from `requestAnimationFrame` to `setTimeout`?

**Review by johannes-weber — COMMENTED** [review]
Thanks for the contribution, @spugachev! I've included some thoughts and suggestions.

File: src/button-group/__tests__/button-group.test.tsx [hunk 30]
```diff
@@ -0,0 +1,263 @@
...
+const renderButtonGroup = (props: Partial<ButtonGroupProps>, ref?: React.Ref<ButtonGroupProps.Ref>) => {
+  const renderResult = render(<ButtonGroup ref={ref} {...defaultProps} {...props} />);
+  return createWrapper(renderResult.container).findButtonGroup()!;
+};
+
+const timeout = async (duration = 0) => {
+  await new Promise(resolve => setTimeout(resolve, duration));
+};
+
+const items1: ButtonGroupProps.ItemOrGroup[] = [
```
[the remark this answers is not in the document]
Comment (spugachev) [reply src/button-group/__tests__/button-group.test.tsx:38]: Changed it to "items"

File: src/button-group/__tests__/button-group.test.tsx [hunk 31]
```diff
@@ -0,0 +1,253 @@
...
+    items: [],
+  },
+];
+
+test('renders all items', () => {
+  const wrapper = renderButtonGroup({ items: items1 });
+
+  expect(wrapper.findItems()).toHaveLength(6);
+  expect(wrapper.findButtonById('edit')).not.toBe(null);
+  expect(wrapper.findMenuById('misc')).not.toBe(null);
```
[the remark this answers is not in the document]
Comment (spugachev) [reply src/button-group/__tests__/button-group.test.tsx:81]: Yes, it would be enough. Removed checks

File: src/internal/hooks/use-focus-visible-state/index.ts (hunk 28 above)
Comment (spugachev) [reply src/internal/hooks/use-focus-visible-state/index.ts:7]: We are not using this anymore, removed the hook

File: src/button-group/styles.scss [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+  display: inline-block;
+}
+
+.item {
+  /* used in test-utils */
+}
+
+.tooltip {
+  /* used in test-utils */
+}
```
[the remark this answers is not in the document]
Comment (spugachev) [reply src/button-group/styles.scss:27]: Moved

File: src/internal/context/single-tab-stop-navigation-context.tsx (hunk 29 above)
Comment (pan-kot) [reply src/internal/context/single-tab-stop-navigation-context.tsx:100]: That is due to Wait for unmounted node to get removed from the DOM comment. The requestAnimationFrame fires before the element is removed and the focus is casted on the element which is then removed.

The code is used in button group, table, and tabs. Using the set timeout does not affect the other component but using the animation frame in button group will fail an integration test designed to capture this behavior.

File: src/button/internal.tsx [hunk 33]
```diff
@@ -45,6 +45,7 @@ export type InternalButtonProps = Omit<ButtonProps, 'variant'> & {
   __iconClass?: string;
   __focusable?: boolean;
   __injectAnalyticsComponentMetadata?: boolean;
+  __title?: string;
```
[the remark this answers is not in the document]
Comment (spugachev) [reply src/button/internal.tsx:48]: We already have a tooltip. Adding a title creates a second tooltip. We want this title to always be empty.
Screenshot 2024-07-16 at 7 50 25 PM

File: pages/button-group/item-permutations.page.tsx [hunk 34]
```diff
@@ -0,0 +1,100 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import React from 'react';
+
+import { ButtonGroup, ButtonGroupProps, SpaceBetween, StatusIndicator } from '~components';
```
[the remark this answers is not in the document]
Comment (spugachev) [reply pages/button-group/item-permutations.page.tsx]: Done

File: pages/button-group/permutations.page.tsx [hunk 35]
```diff
@@ -0,0 +1,106 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import React from 'react';
+
+import { ButtonGroup, ButtonGroupProps } from '~components';
```
[the remark this answers is not in the document]
Comment (spugachev) [reply pages/button-group/permutations.page.tsx]: Done

File: src/internal/context/single-tab-stop-navigation-context.tsx (hunk 29 above)
Comment (johannes-weber) [reply src/internal/context/single-tab-stop-navigation-context.tsx:100]: Got it, thanks for clarification!

File: src/button-group/__tests__/button-group.test.tsx (hunk 21 above)
Comment (spugachev) [reply src/button-group/__tests__/button-group.test.tsx]: Don't need it anymore. Removed

File: src/button-group/icon-button-item.tsx (hunk 22 above)
Comment (spugachev) [reply src/button-group/icon-button-item.tsx:46]: Good question. It can be iconName, iconAlt or iconSvg. We can't make iconName requiried. 
@pan-kot waht do you think, should we display nothing if no icon defined?

File: src/button-group/interfaces.ts (hunk 23 above)
Comment (spugachev) [reply src/button-group/interfaces.ts:12]: I cahnged it to:
  /**
   * Adds `aria-label` to the button group toolbar element.
   * Use this to provide a unique accessible name for buttons groups on the page.
   */

File: src/button-group/menu-dropdown-item.tsx (hunk 27 above)
Comment (pan-kot) [reply src/button-group/menu-dropdown-item.tsx:64]: Do you see a risk with that?

These test-IDs are always used in combination with class names. We need the one on the button-dropdown so that the test-utils elector can point to the component's root so that we can wrap it with ButtonDropdownWrapper.

The second one is used in the internal.tsx to find the focus targets that are the actual focusable elements so we need it to point to the trigger. We can technically find the button dropdown and use a nested selector to find a trigger but that would increase those queries complexity by much as we'd need to assert item types to figure out the selector to use.

File: src/button-group/menu-dropdown-item.tsx (hunk 26 above)
Comment (spugachev) [reply src/button-group/menu-dropdown-item.tsx]: Extracteed MenuDropdownItemProps to interfaces.ts

File: src/button-group/item-element.tsx (hunk 25 above)
Comment (spugachev) [reply src/button-group/item-element.tsx]: Extracted ItemElementProps to interfaces.ts

File: src/button-group/menu-dropdown-item.tsx (hunk 27 above)
Comment (spugachev) [reply src/button-group/menu-dropdown-item.tsx]: @pan-kot I think we can keep only one. Am I right?

File: src/button-group/__tests__/button-group.test.tsx [hunk 36]
```diff
@@ -0,0 +1,253 @@
...
+  const onItemClick = jest.fn();
+  const wrapper = renderButtonGroup({ items: items1, onItemClick });
+  const buttonDropdown = wrapper.findMenuById('misc')!;
+  buttonDropdown.openDropdown();
+  buttonDropdown.findItemById('menu-open')!.click();
+
+  expect(onItemClick).toHaveBeenCalledWith(expect.objectContaining({ detail: { id: 'menu-open' } }));
+});
+
+describe('focus', () => {
```
[the remark this answers is not in the document]
Comment (spugachev) [reply src/button-group/__tests__/button-group.test.tsx:95]: Added a test that focus returned to the dropdown button

File: src/button-group/icon-button-item.tsx (hunk 22 above)
Comment (johannes-weber) [reply src/button-group/icon-button-item.tsx:46]: What about a `warnOnce` if none of them is provided?

Comment (spugachev) [reply src/button-group/icon-button-item.tsx:46]: Yes, we already have warnOnce

File: src/button-group/menu-dropdown-item.tsx (hunk 27 above)
Comment (pan-kot) [reply src/button-group/menu-dropdown-item.tsx]: Addressed here: https://github.com/cloudscape-design/components/pull/2469/commits/4a40ee45296b5b217aeb03a14abe0156f51feaab

**Review by johannes-weber — APPROVED** [review]

**Review by pan-kot — DISMISSED** [review]
Great work @spugachev !

**Review by pan-kot — APPROVED** [review]

**Review by pan-kot — APPROVED** [review]
